File: services/ai/app/rules_engine.py
"""Deterministic IMNCI rule evaluator.

Loads the versioned rule table from rules/imnci-rules.v1.json (single source
of truth, shared with the TypeScript on-device evaluator in apps/web) and
decides severity from a facts object. This module is the ONLY place severity
is decided in this service -- the LLM extraction pipeline must never be
allowed to set severity itself. See CONTRACT.md.
"""
from __future__ import annotations

import json
from functools import lru_cache
from pathlib import Path
from typing import Any

from . import config

SEVERITY_RANK = {"GREEN": 0, "YELLOW": 1, "RED": 2}


class RulesFileNotFound(FileNotFoundError):
    pass


def _candidate_paths() -> list[Path]:
    here = Path(__file__).resolve()
    candidates = []
    if config.RULES_PATH:
        candidates.append(Path(config.RULES_PATH))
    # services/ai/app/rules_engine.py -> parents: app, ai, services, repo_root
    candidates.append(here.parents[3] / "rules" / "imnci-rules.v1.json")
    # Docker image layout: rules/ copied alongside the service (services/ai/rules)
    candidates.append(here.parents[1] / "rules" / "imnci-rules.v1.json")
    candidates.append(Path.cwd() / "rules" / "imnci-rules.v1.json")
    return candidates


def resolve_rules_path() -> Path:
    for candidate in _candidate_paths():
        if candidate.is_file():
            return candidate
    raise RulesFileNotFound(
        "Could not locate imnci-rules.v1.json. Tried: "
        + ", ".join(str(c) for c in _candidate_paths())
        + ". Set RULES_PATH to override."
    )


@lru_cache(maxsize=1)
def load_rules() -> dict[str, Any]:
    path = resolve_rules_path()
    with open(path, "r", encoding="utf-8") as f:
        return json.load(f)


def clear_cache() -> None:
    load_rules.cache_clear()


def _get_field(facts: dict[str, Any], field: str) -> Any:
    return facts.get(field, None)


def _eval_leaf(cond: dict[str, Any], facts: dict[str, Any]) -> bool:
    field = cond["field"]
    op = cond["op"]
    value = _get_field(facts, field)

    if op == "exists":
        return value is not None

    # Per rules/imnci-rules.v1.json condition_grammar.missing_field_semantics:
    # a missing/null fact never satisfies a leaf condition, regardless of op.
    if value is None:
        return False

    target = cond.get("value")
    if op == "eq":
        return value == target
    if op == "neq":
        return value != target
    if op in ("gt", "gte", "lt", "lte"):
        if not isinstance(value, (int, float)) or isinstance(value, bool):
            return False
        if not isinstance(target, (int, float)) or isinstance(target, bool):
            return False
        if op == "gt":
            return value > target
        if op == "gte":
            return value >= target
        if op == "lt":
            return value < target
        if op == "lte":
            return value <= target
    raise ValueError(f"Unknown condition op: {op!r}")


def eval_condition(cond: dict[str, Any], facts: dict[str, Any]) -> bool:
    if "all" in cond:
        return all(eval_condition(c, facts) for c in cond["all"])
    if "any" in cond:
        return any(eval_condition(c, facts) for c in cond["any"])
    if "atLeast" in cond:
        spec = cond["atLeast"]
        n = spec["n"]
        count = sum(1 for c in spec["conditions"] if eval_condition(c, facts))
        return count >= n
    if "field" in cond:
        return _eval_leaf(cond, facts)
    raise ValueError(f"Malformed condition, no recognized key: {cond!r}")


def evaluate(facts: dict[str, Any], rules_table: dict[str, Any] | None = None) -> dict[str, Any]:
    """Return {severity, rule_id, rule_version, matched_rules} for a facts object.

    severity = highest-ranked severity among every rule whose `when` matches;
    falls back to the table's default_rule_id/default_severity if none match.
    """
    table = rules_table if rules_table is not None else load_rules()

    matched = [rule for rule in table["rules"] if eval_condition(rule["when"], facts)]

    if not matched:
        return {
            "severity": table["default_severity"],
            "rule_id": table["default_rule_id"],
            "rule_version": table["version"],
            "matched_rules": [],
        }

    best = max(matched, key=lambda r: SEVERITY_RANK[r["severity"]])
    return {
        "severity": best["severity"],
        "rule_id": best["id"],
        "rule_version": table["version"],
        "matched_rules": [r["id"] for r in matched],
    }

File: apps/web/lib/rules-engine.ts
// On-device implementation of the IMNCI rule table grammar defined in
// rules/imnci-rules.v1.json (`condition_grammar`). This is the offline
// fallback evaluator used by the ASHA app's structured checklist per
// CLAUDE.md's offline-degradation rule — it must stay behaviorally
// identical to the Python evaluator in services/ai. No Node-only APIs:
// this file runs in the browser (on-device severity calc) as well as in
// Next.js route handlers.
import ruleTable from "./imnci-rules.v1.json";
import type { ImnciFacts, Severity, TriageResult } from "./types";

type Op = "eq" | "neq" | "gt" | "gte" | "lt" | "lte" | "exists";

interface LeafCondition {
  field: string;
  op: Op;
  value?: unknown;
}

interface AllCondition {
  all: Condition[];
}

interface AnyCondition {
  any: Condition[];
}

interface AtLeastCondition {
  atLeast: { n: number; conditions: Condition[] };
}

type Condition = LeafCondition | AllCondition | AnyCondition | AtLeastCondition;

interface Rule {
  id: string;
  category: string;
  severity: Severity;
  description: string;
  when: Condition;
}

interface RuleTable {
  rule_table_id: string;
  version: string;
  severity_order: Severity[];
  default_severity: Severity;
  default_rule_id: string;
  default_description: string;
  rules: Rule[];
}

const table = ruleTable as unknown as RuleTable;

function getField(facts: ImnciFacts, field: string): unknown {
  const value = (facts as Record<string, unknown>)[field];
  return value === undefined ? null : value;
}

function evalLeaf(facts: ImnciFacts, leaf: LeafCondition): boolean {
  const actual = getField(facts, leaf.field);

  if (leaf.op === "exists") {
    return actual !== null;
  }

  if (actual === null) {
    // Missing field: booleans/equality never auto-fire, numeric comparisons
    // are always false. Fail-safe default per CONTRACT.md.
    return false;
  }

  switch (leaf.op) {
    case "eq":
      return actual === leaf.value;
    case "neq":
      return actual !== leaf.value;
    case "gt":
      return typeof actual === "number" && typeof leaf.value === "number" && actual > leaf.value;
    case "gte":
      return typeof actual === "number" && typeof leaf.value === "number" && actual >= leaf.value;
    case "lt":
      return typeof actual === "number" && typeof leaf.value === "number" && actual < leaf.value;
    case "lte":
      return typeof actual === "number" && typeof leaf.value === "number" && actual <= leaf.value;
    default:
      return false;
  }
}

function evalCondition(facts: ImnciFacts, condition: Condition): boolean {
  if ("all" in condition) {
    return condition.all.every((c) => evalCondition(facts, c));
  }
  if ("any" in condition) {
    return condition.any.some((c) => evalCondition(facts, c));
  }
  if ("atLeast" in condition) {
    const matched = condition.atLeast.conditions.filter((c) => evalCondition(facts, c)).length;
    return matched >= condition.atLeast.n;
  }
  return evalLeaf(facts, condition as LeafCondition);
}

function severityRank(severity: Severity): number {
  return table.severity_order.indexOf(severity);
}

/**
 * Evaluates every rule in the table against the given facts and returns the
 * highest-severity match (RED > YELLOW > GREEN). Falls back to the table's
 * default rule/severity when nothing matches. Deterministic and pure — same
 * facts always produce the same result, which is what makes each decision
 * auditable.
 */
export function evaluateTriage(facts: ImnciFacts): TriageResult {
  let best: { rule: Rule } | null = null;

  for (const rule of table.rules) {
    if (!evalCondition(facts, rule.when)) continue;
    if (!best || severityRank(rule.severity) > severityRank(best.rule.severity)) {
      best = { rule };
    }
  }

  if (!best) {
    return {
      severity: table.default_severity,
      rule_id: table.default_rule_id,
      rule_version: table.version,
      matched_description: table.default_description,
    };
  }

  return {
    severity: best.rule.severity,
    rule_id: best.rule.id,
    rule_version: table.version,
    matched_description: best.rule.description,
  };
}

export function getRuleTableVersion(): string {
  return table.version;
}

/** Looks up a rule's description by id (including the default rule) — used to render a consistent explanation regardless of whether severity came from the LLM path or the on-device checklist path. */
export function getRuleDescription(ruleId: string): string {
  if (ruleId === table.default_rule_id) return table.default_description;
  return table.rules.find((r) => r.id === ruleId)?.description ?? "";
}

export function getFactFieldNames(): string[] {
  return Object.keys((ruleTable as { fact_schema: Record<string, unknown> }).fact_schema);
}
